Devtools

maxentile opened the pull request:
Standard structure, following John's comment: https://github.com/choderalab/msm-pipeline/pull/9#issuecomment-234250613

maxentile committed:
move files into msmpipeline/ directory

maxentile committed:
Move example torque script into cluster-utils/ directory

maxentile committed:
Add setup.py, closely mimicking the FAHMunge setup.py

- Mostly just took this file:
https://github.com/choderalab/fahmunge/blob/master/setup.py and
replaced "FAHmunge" with "msmpipeline"
- Added Sonya Hanson, Steven Albanese, and Josh Fass as authors

maxentile committed:
Add travis-ci install script

Follows example in:
https://github.com/choderalab/fahmunge/blob/master/devtools/travis-ci/in
stall.sh

maxentile committed:
Add conda-recipe/build.sh

maxentile committed:
Add meta.yaml

maxentile committed:
Add versioneer info to setup.cfg

maxentile committed:
Add versioneer info to __init__.py

maxentile committed:
Add versioneer

maxentile committed:
Add requirements.txt

maxentile committed:
Add *.pyc to .gitignore

maxentile committed:
Update setup.py

Based on this template:
https://gist.github.com/juancarlospaco/75d258d6ffb6e15f395c

maxentile committed:
Merge remote-tracking branch 'choderalab/master' into devtools

maxentile committed:
Minimal setup.py

maxentile committed:
Ignore Python distribution / packaging files

maxentile committed:
Oops, forgot .travis.yaml

maxentile committed:
.travis.yaml --> .travis.yml

"Could not find .travis.yml, using standard configuration."

maxentile committed:
Fix url

jchodera commented:
Great start, but you'll need to update the `.travis.yml` to use the testing framework too:

Here's an example:
https://github.com/choderalab/alchemy/blob/master/.travis.yml

We probably want to make your pipeline script an "entry point", where running a script command (like `msm-pipeline`) that gets installed into `~/miniconda/bin/` will call into an entry point like `msmpipeline/cli.py` into a function like `run_pipeline()` or something. More info on entry points here:
- http://python-packaging.readthedocs.io/en/latest/command-line-scripts.html
- https://chriswarrick.com/blog/2014/09/15/python-apps-the-right-way-entry_points-and-scripts/

maxentile committed:
Add entry points

maxentile committed:
Fix .travis.yml

maxentile committed:
Fix .travis.yml?

jchodera commented:
Why not use `conda` to install `omnia` dependencies? Are you using things that aren't available through `omnia`?

maxentile committed:
Ignore test output

maxentile committed:
Remove entry_points from meta.yaml

- Moved to setup.py

maxentile committed:
Delete requirements.txt

- Conda exists

maxentile committed:
Update Conda recipe

maxentile committed:
Simplify

- Moved "python setup.py install" from build.sh into a line of meta.yaml
- Replace travis-ci boilerplate (insteall miniconda) with 3 lines in
.travis.yml
- Turn off Travis email notifications!

maxentile committed:
Fix license field in setup.py

- Was inconsistent with license field in conda-recipe/meta.yaml

maxentile committed:
Install conda-build

maxentile committed:
Yes, please install conda-build

maxentile committed:
Temporarily remove pip-installed dependency

- Causes a mysterious Travis error

maxentile committed:
Add omnia channel

maxentile committed:
Oops, typo

maxentile committed:
Specify same entry points in meta.yaml as setup.py

maxentile committed:
Ignore test-report entry point

- Not sure why "test-report" isn't a valid entry point during the
Travis test -- I think it's specified correctly in both setup.py and
meta.yaml...

maxentile committed:
Move build requirements from "run" to "build"

maxentile committed:
Fix travis?

Travis error: "ImportError: No module named
msmbuilder.example_datasets" even though Travis logs confirm msmbuilder
was installed.

Hypothesis: requirements are being installed to a different conda
environment than is being tested

Attempted solution: switch to testing environment before installation

maxentile committed:
Fix Travis?

- Previous Travis log shows that "source activate
/home/travis/miniconda/envs/_test" was executed, but including that
command in before_install leads to an error
- Create and switch to test-environment instead
- Also have conda log more information

maxentile committed:
Oops, use correct signature for conda create

maxentile committed:
Don't have conda wait for a non-existent user to type 'y'

maxentile committed:
Move matplotlib.pyplot import

Attempts to address RuntimeError: "Invalid DISPLAY variable"

Based on this helpful UserWarning: "This call to matplotlib.use() has
no effect because the backend has already been chosen; matplotlib.use()
must be called *before* pylab, matplotlib.pyplot, or
matplotlib.backends is imported for the first time."

maxentile committed:
Use test-report entry point

maxentile committed:
Avoid a time-consuming download

- Replace MetEnkaphalin example with random data
- Also apparently trim extraneous whitespace. Thank you, Atom!

maxentile committed:
Move corner import to beginning of generate_report

maxentile committed:
.jpg --> .png

maxentile committed:
Reintroduce pip-installed dependency

maxentile committed:
Ignore .png format test outputs as well

maxentile committed:
Make sure matplotlib.use('Agg') is called before pyplot import

Still getting "UserWarning:  This call to matplotlib.use() has no
effect because the backend has already been chosen; matplotlib.use()
must be called *before* pylab, matplotlib.pyplot, or
matplotlib.backends is imported for the first time."

maxentile committed:
Okay, apparently can't use test-report entry point

maxentile committed:
Oops, correct path to test script

maxentile committed:
Add quick_test

Ignore plotting scripts, just make sure we can generate and analyze a
test instance...

maxentile committed:
Oops, had specified Python versions twice

Results in redundant tests on Travis

maxentile committed:
Update README.md

- Add Travis build status badge
- Use 'msm-pipeline' entry point instead of 'python pipeline.py'

maxentile committed:
Include python setup.py install again

Maybe this will get the entry points to register in Travis?

maxentile committed:
Did that make the entry points visible to Travis?

maxentile committed:
Oops, needed to specify a function

jchodera commented:
Sorry for all the travis pains.

We've tried to minimize these by sticking to a fairly standard structure for most of our projects.

For example, in [`smarty`](https://github.com/open-forcefield-group/smarty/blob/master/.travis.yml), we have this block:

``` YAML
language: c
sudo: false

branches:
  only:
    - master

install:
  - source devtools/travis-ci/install.sh
  - export PYTHONUNBUFFERED=true

script:
  # Add omnia channel
  - conda config --add channels ${ORGNAME}
  # build the recipe
  - conda build devtools/conda-recipe
  # Test the local installation
  - source activate _test
  # This is a temporary workaround for the conda issues.
  - conda install --yes nose nose-timer numpy networkx lxml openmm
  #- pip install .
  # Install OpenEye toolkit
  - pip install $OPENEYE_CHANNEL openeye-toolkits && python -c "import openeye; print(openeye.__version__)"
  # Run tests
  #- cd devtools && nosetests $PACKAGENAME --nocapture --verbosity=2 --with-doctest --with-timer -a "\!slow" && cd ..
  - python --version
  - cd devtools && nosetests -vv --nocapture --with-timer $PACKAGENAME && cd ..

env:
  matrix:
    - python=2.7  CONDA_PY=27
    - python=3.4  CONDA_PY=34
    - python=3.5  CONDA_PY=35

  global:
    - ORGNAME="omnia"
    - PACKAGENAME="smarty"
```

The `install:` section installs Miniconda, while the `script:` section builds the conda dev recipe in `devtools/conda-recipe`. If you have a `test:` section in the conda dev recipe, it also sets up a conda environment called `_test` that you can also install other stuff into (as we do here, with the OpenEye tools). You can test your script inside the `conda-recipe/meta.yaml`, as we do [here](https://github.com/choderalab/yank/blob/master/devtools/conda-recipe/meta.yaml#L55-L63) for YANK:

``` YAML
test:
  requires:
    - nose
    - nose-timer
  imports:
    - yank
  commands:
    - yank --help
    - yank selftest
```

maxentile commented:
Thanks, John!

And sorry to all subscribers for the deluge of Github notification spam the last few hours -- most of this complication was self-inflicted. I first directly copied this structure from FAHMunge, and then I tried to simplify / remove bits of the "boilerplate" that didn't seem needed here (e.g. pushing docs to S3) and inadvertently broke a bunch of things, which I think are now mostly fixed:
- [setup.py](https://github.com/choderalab/fahmunge/blob/master/setup.py) -- Removed almost all of this, retaining minimal package information and the call to `versioneer.get_version()`
- [devtools/travis-ci/](https://github.com/choderalab/fahmunge/tree/master/devtools/travis-ci)-- The only thing I think I need to retain from these scripts is to fetch and install Miniconda -- I replaced this with a few lines in `.travis.yml`.
- [devtools/conda-recipe/](https://github.com/choderalab/fahmunge/tree/master/devtools/conda-recipe) -- Moved the `build.sh` one-liner into `meta.yaml`. Also didn't duplicate the readme.

Remaining minutia: 
1. `versioneer.py` sets the version in `setup.py`, but the version is also hardcoded in `meta.yaml`. To-do: use `versioneer`'s version inside the `meta.yaml` file.
2. Set the matplotlib backend correctly. I think I've made sure `matplotlib.use('Agg')` is before every use of matplotlib, but I keep getting an "Invalid DISPLAY variable error" accompanied by this warning: "This call to matplotlib.use() has no effect because the backend has already been chosen; matplotlib.use() must be called _before_ pylab, matplotlib.pyplot, or matplotlib.backends is imported for the first time."

jchodera commented:
The reason we avoid mucking about with the stock boilerplate is to avoid
the technical debt of having everything be different. By spending a day of
self-inflicted pain as a direct result of mucking with it, I hope you
understand now why we don't do this. :)

For (1), you can't use the output of versioneer in meta.yaml, to my
knowledge. @jhprinz has looked into this, but there may bit be a good
solution yet.

For (2), I figured this out in some of my scripts --- trying to figure out
which ones.

jchodera commented:
I think this idiom should work for the [matplotlib issue](https://github.com/choderalab/perses/blob/d6907868fe69030a8c11263b28b3580190e4a275/perses/analysis/analysis.py#L24-L29):

``` python
import matplotlib as mpl
mpl.use('Agg')
import seaborn as sns

from matplotlib.backends.backend_pdf import PdfPages
import matplotlib.pyplot as plt
```

maxentile committed:
Fix matplotlib issue?

Try also importing seaborn and the PdfPages backend, as per:
https://github.com/choderalab/msm-pipeline/pull/19#issuecomment-24030591
7

maxentile commented:
Thanks John! I'll try that now.

jchodera commented on .travis.yml:
```diff
@@ -0,0 +1,30 @@
+install:
+  - conda create --name test-environment python=$TRAVIS_PYTHON_VERSION pyemma msmbuilder
+  - source activate test-environment
+  - conda build devtools/conda-recipe
```
This doesn't actually install any of your dependencies in the `test-environment` before calling `python setup.py install`, so everything will fail.

I'd highly recommend using the idiom we've settled on:

``` YAML
script:
  # Add omnia channel
  - conda config --add channels ${ORGNAME}
  # Add omnia dev channel
  - conda config --add channels https://conda.anaconda.org/omnia/label/dev
  # build the recipe
  - conda build devtools/conda-recipe
  # Test the local installation
  - source activate _test
  # Run tests
  - cd devtools && nosetests $PACKAGENAME --nocapture --verbosity=2 --with-doctest --with-timer && cd ..
```

The `source activate _test` will use the environment set up by the `test:` section of your `conda build` call, which also includes your `run:` dependencies.

maxentile committed:
Remove unnecessary stuff, use correct devtools setup

maxentile committed:
Fetch data from msmbuilder.example_datasets instead of storing a chunk of it

- Also use sudo on travis, probably avoids permission error?

maxentile commented:
Will sort out the Travis error later. Merging.
